src/compiler/program.ts

```diff
@@ -285,7 +285,7 @@ namespace ts {
         return resolutions;
     }
 
-    export function createProgram(rootNames: string[], options: CompilerOptions, host?: CompilerHost, oldProgram?: Program): Program {
+    export function createProgram(rootNames: string[], options: CompilerOptions, host?: CompilerHost, oldProgram?: Program, fileExtensionMap?: FileExtensionMap): Program {

```

Comment:
I don't think we need to push concrete knowledge about these extra extensions through all layers, this will be necessary if we'll need to be able to add files with these extensions into program via imports or tripleslash references. Since both these scenarios are not supported and files with custom extensions should always be included in the list of root files I think we can:
- revert changes in program, language service host, lsHost and services
- in editor services for configured projects if host configuration specifies extra extensions and result of cracking config files contains files with these extensions - set `allowNonTsExtensions` bit on `CompilerOptions` to force adding these files into program.

this way I think we can update only project system part and keep all other layers relatively untouched.